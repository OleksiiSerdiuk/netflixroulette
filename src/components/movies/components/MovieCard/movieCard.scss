.movie-card {
  width: 100%;
  position: relative;

  &__img {
    width: 100%;
    height: auto;

    &-wrapper {
      background: rgb(204 193 193 / 50%);
      min-height: 480px;
      color: white;
      text-decoration: inherit;
      display: flex;
      align-items: center;
      text-align: center;
    }
  }

  &__body {
    margin-top: 10px;
  }

  &__header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-bottom: 5px;
  }

  &__genres {
    font-weight: 500;
    font-size: 14px;
    line-height: 17px;
  }

  &__title {
    font-weight: 500;
    font-size: 18px;
    line-height: 22px;
  }

  &__date {
    border: 1px solid var(--color-dusty-gray);
    border-radius: 4px;
    font-weight: 500;
    font-size: 14px;
    line-height: 17px;
    padding: 5px 10px;
  }

  &__settings {
    position: absolute;
    right: 0;
    top: 0;
    width: 100%;
    display: flex;
    justify-content: right;
    z-index: 2;
  }

  &__menu {
    width: 100%;
    max-width: 190px;
    background: var(--color-mine-shaft);
    padding-top: 35px;
    padding-bottom: 10px;

    li {
      font-weight: 400;
      font-size: 16px;
      line-height: 20px;
      padding: 10px 25px;
      cursor: pointer;
      transition: all 0.3s;

      &:hover,
      &:focus {
        background: var(--color-carnation);
      }
    }
  }

  &__icon {
    cursor: pointer;
    transition: 0.3s;
    position: absolute;
    top: 0;
    right: 0;
    z-index: 5;

    &:hover,
    &:focus {
      opacity: 0.8;
    }

    .icon-close {
      width: 24px;
      height: 24px;

      &:before,
      &:after {
        height: 22px;
        left: 8px;
        top: 5px;
      }
    }
  }
}
