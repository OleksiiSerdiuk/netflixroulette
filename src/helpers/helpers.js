export const parseDate = (date) => {
  if (!date) return;

  return new Date(date).getFullYear();
};

export const joinArrayBySeparator = (arr, separator = ', ') => {
  if (arr.length === 0) return;

  return arr.join(separator);
};

export const getTimeTitle = (currentValue) => {
  const quotient = Math.floor(currentValue / 60);
  const remainder = currentValue % 60;

  if (currentValue > 60) {
    return `${quotient}h ${remainder !== 0 && `${remainder}min`}`;
  } else {
    return `${currentValue} min`;
  }
};
