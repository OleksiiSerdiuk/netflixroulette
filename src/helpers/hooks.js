import { useState } from 'react';

export const useLoader = () => {
  const [loading, setLoading] = useState(false);

  return [
    loading ? <p className='color-white'>loading...</p> : null,
    () => setLoading(true),
    () => setLoading(false),
  ];
};
