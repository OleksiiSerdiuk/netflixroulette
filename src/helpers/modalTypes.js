export const MODAL_TYPES_MAP = {
  ADD: 'add',
  EDIT: 'edit',
  DELETE: 'delete',
  CONGRATULATIONS: 'congratulations',
};
