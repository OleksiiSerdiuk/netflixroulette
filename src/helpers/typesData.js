import PropTypes from 'prop-types';

export const movieTypes = PropTypes.shape({
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  title: PropTypes.string.isRequired,
  tagline: PropTypes.string,
  vote_average: PropTypes.number,
  vote_count: PropTypes.number,
  release_date: PropTypes.string,
  poster_path: PropTypes.string.isRequired,
  overview: PropTypes.string.isRequired,
  budget: PropTypes.number,
  revenue: PropTypes.number,
  runtime: PropTypes.number.isRequired,
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
});
