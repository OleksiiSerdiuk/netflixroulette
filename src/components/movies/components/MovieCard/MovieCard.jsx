import './movieCard.scss';
import { parseDate, joinArrayBySeparator } from '../../../../helpers/helpers';
import { useState, useContext } from 'react';
import { ModalContext } from '../../../../contexts';
import ToggleImg from './../../../../images/toggle.svg';
import { mockMenu } from '../../../../helpers/mockData';
import { MODAL_TYPES_MAP } from '../../../../helpers/modalTypes';
import { movieTypes } from '../../../../helpers/typesData';
import { MovieContext } from '../../../../contexts';
import { Link } from 'react-router-dom';

const MovieCard = ({ movie }) => {
  const { id, title, poster_path, release_date, genres } = movie;
  const { openModal } = useContext(ModalContext);
  const { setMovieDescription } = useContext(MovieContext);

  const [isShowMenu, setIsShowMenu] = useState(false);

  const movieReleaseDate = parseDate(release_date) || 'empty';
  const movieGenres = joinArrayBySeparator(genres) || 'empty';

  const toggleMenu = () => {
    setIsShowMenu(!isShowMenu);
  };

  const handlerMenu = (menuItemId) => {
    const isEdit = menuItemId === MODAL_TYPES_MAP.EDIT;
    const modalType = isEdit ? MODAL_TYPES_MAP.EDIT : MODAL_TYPES_MAP.DELETE;
    const payload = isEdit ? movie : movie.id;

    toggleMenu();
    openModal({ modalType, payload });
  };

  const showMovie = (movie) => {
    setMovieDescription(movie);
  };

  return (
    <div className='movie-card' id={id}>
      <div className='movie-card__settings'>
        <div
          className={`movie-card__icon ${isShowMenu && 'movie-card__icon'}`}
          onClick={toggleMenu}
          aria-hidden='true'
        >
          {isShowMenu ? (
            <span className='icon-close' />
          ) : (
            <img src={ToggleImg} alt='Toggle' />
          )}
        </div>

        {isShowMenu && (
          <ul className='movie-card__menu'>
            {mockMenu.map(({ id, title }) => {
              return (
                <li key={id} onClick={() => handlerMenu(id)} aria-hidden='true'>
                  {title}
                </li>
              );
            })}
          </ul>
        )}
      </div>

      <Link to={`/movie/:${id}`} type='button' onClick={() => showMovie(movie)}>
        <div className='movie-card__img-wrapper'>
          <img className='movie-card__img' src={poster_path} alt={title} />
        </div>
      </Link>

      <div className='movie-card__body'>
        <div className='movie-card__header'>
          <h2 className='movie-card__title'>{title}</h2>
          <span className='movie-card__date'>{movieReleaseDate}</span>
        </div>
        <p className='movie-card__genres'>{movieGenres}</p>
      </div>
    </div>
  );
};

const propTypes = {
  movie: movieTypes,
};

const defaultProps = {
  movie: {},
};

MovieCard.propTypes = propTypes;
MovieCard.defaultProps = defaultProps;

export default MovieCard;
