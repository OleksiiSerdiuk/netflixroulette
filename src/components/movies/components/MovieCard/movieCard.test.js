import { render, screen } from '@testing-library/react';
import MovieCard from './MovieCard';
import { mockMovie } from '../../../../helpers/mockData';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';

describe('movie card', () => {
  test('renders movie card snapshot', () => {
    const { asFragment, container } = render(
      <BrowserRouter>
        <MovieCard movie={mockMovie} />
      </BrowserRouter>
    );
    expect(asFragment(<MovieCard movie={mockMovie} />)).toMatchSnapshot();
    expect(screen.getByText(mockMovie.title)).toBeInTheDocument();
    const getImgTag = container
      .querySelector('a')
      .querySelector('img')
      .getAttribute('src');
    expect(getImgTag).toBe(mockMovie.poster_path);
  });
});
