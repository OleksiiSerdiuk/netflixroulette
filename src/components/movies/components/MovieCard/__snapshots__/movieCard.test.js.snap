// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movie card renders movie card snapshot 1`] = `
<DocumentFragment>
  <div
    class="movie-card"
    id="337167"
  >
    <div
      class="movie-card__settings"
    >
      <div
        aria-hidden="true"
        class="movie-card__icon false"
      >
        <img
          alt="Toggle"
          src="test-file-stub"
        />
      </div>
    </div>
    <a
      href="/movie/:337167"
      type="button"
    >
      <div
        class="movie-card__img-wrapper"
      >
        <img
          alt="Fifty Shades Freed"
          class="movie-card__img"
          src="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
        />
      </div>
    </a>
    <div
      class="movie-card__body"
    >
      <div
        class="movie-card__header"
      >
        <h2
          class="movie-card__title"
        >
          Fifty Shades Freed
        </h2>
        <span
          class="movie-card__date"
        >
          2018
        </span>
      </div>
      <p
        class="movie-card__genres"
      >
        Drama, Romance
      </p>
    </div>
  </div>
</DocumentFragment>
`;
